File: dsl/disjoint_set.cpp
/* DSL1_A: Disjoint Set */

// Solve using Union-Find Tree

#include <iostream>
#include <cstdio>

int *parents;
int *ranks;

int find(int x) {
    if (parents[x] == x)
        return x;
    else
        return parents[x] = find( parents[x] );
}

void unite(int x, int y) {
    x = find(x);
    y = find(y);
    if ( ranks[x] < ranks[y] ) {
        parents[x] = y;
    }
    else if ( ranks[y] < ranks[x] ) {
        parents[y] = x;
    }
    else if (x != y) {
        parents[x] = y;
        ranks[y] += 1;
    }
}

int main() {
    int n, q;
    scanf("%d %d", &n, &q);

    parents = new int[n];
    ranks   = new int[n];
    for (int i = 0; i < n; i++) {
        parents[i] = i;
        ranks[i]   = 0;
    }

    int command, x, y;
    for (int i = 0; i < q; i++) {
        scanf("%d %d %d\n", &command, &x, &y);
        switch (command) {
            case 0:
                unite(x, y);
                break;
            case 1:
                printf("%d\n", find(x) == find(y));
                break;
        }
    }

    return 0;
}

File: dsl/disjoint_set.rb
# DSL1_A: Disjoint Set


class DisjointSet
  def initialize(size)
    @parents = size.times.map { |i| i }
    @sizes   = Array.new(size, 0)
  end

  def unite_set(x, y)
    x = find_root(x)
    y = find_root(y)
    sub, main = @sizes[x] < @sizes[y] ? [x, y] : [y, x]
    @sizes[main] += 1 if @sizes[main] == @sizes[sub]
    @parents[sub] = main
  end

  def same_set?(x, y)
    find_root(x) == find_root(y)
  end

  def find_root(x)
    unless @parents[x] == x
      @parents[x] = find_root @parents[x]
    end
    @parents[x]
  end
end

module Query
  UNITE = 0
  SAME  = 1
end

size, n_queries = gets.split.map &:to_i
djset = DisjointSet.new(size)

n_queries.times do
  query, x, y = gets.split.map &:to_i
  case query
  when Query::UNITE
    djset.unite_set(x, y)
  when Query::SAME
    puts djset.same_set?(x, y) ? 1 : 0
  end
end

